bittensor/axon.py

```diff
@@ -0,0 +1,452 @@
+            Exception: If the nonce is not larger than the previous nonce for the same endpoint key.
+            Exception: If the signature verification fails.
+
+        After successful verification, the nonce for the given endpoint key is updated.
+
+        Note:
+            The verification process assumes the use of an asymmetric encryption algorithm,
+            where the sender signs the message with their private key and the receiver verifies the signature using the sender's public key.
+        """
+        # Build the keypair from the dendrite_hotkey
+        keypair = Keypair(ss58_address = headers.dendrite_hotkey)
+
+        # Build the signature messages.
+        message = f"{headers.dendrite_nonce}.{headers.dendrite_hotkey}.{self.wallet.hotkey.ss58_address}.{headers.dendrite_uuid}"
+
+        # Build the unique endpoint key.
+        endpoint_key = f"{headers.dendrite_hotkey}:{headers.dendrite_uuid}"
+
+        # Check the nonce from the endpoint key.
+        if endpoint_key in self.nonces.keys():
+
+            # Ensure the nonce increases.
+            if headers.dendrite_nonce <= self.nonces[endpoint_key]:
+                raise Exception("Nonce is too small")
+            
+        if not keypair.verify(message, headers.dendrite_signature):
+            raise Exception("Signature mismatch")
+        
+        # Success
+        self.nonces[endpoint_key] = headers.dendrite_nonce
+    
+
+class AxonMiddleware(BaseHTTPMiddleware):
+    def __init__(self, app, axon ):
+        super().__init__(app)
+        self.axon = axon        
+
+    async def dispatch(self, request: Request, call_next: RequestResponseEndpoint) -> Request:
+
+        bittensor.logging.trace('enter dispatch')
+        start_time = time.time()
+        headers = bittensor.Headers( **dict(request.headers) )
+        headers.__dict__.update({
+            'request_name': str(request.url.path.split("/")[1]),
+            'axon_version': str(bittensor.__version_as_int__),
+            'axon_uuid': str(self.axon.uuid),
+            'axon_nonce': f"{time.monotonic_ns()}",
+            'axon_status_message': "Success",
+            'axon_status_code': "100",
+            'axon_process_time': "0",
+            'axon_signature': "",
+            'dendrite_port': str(request.client.port),
+            'dendrite_ip': str(request.client.host),
+        })
+        try:
+            # Log success
+            bittensor.logging.debug( f"axon     | --> | {headers.request_name} | {headers.dendrite_hotkey} | {headers.dendrite_ip}:{headers.dendrite_port} | 200 | Success ")
+
+            # Check verification     
+            bittensor.logging.trace('Check verification')
+            verify_fn = self.axon.verify_fns[ headers.request_name ]
+            if verify_fn:
+                try: verify_fn( headers )
+                except Exception as e:
+                    bittensor.logging.trace(f'Verify exception')
+                    headers.axon_status_code = "401"
```

Comment:
Love the solution to put the status code in the mutable headers. Avoids returning or raising issues with `JSONResponse` and `StreamingReponse` objects. 👍